### PR TITLE
docs: switch dspy example to weaviate and cohere

### PR DESCRIPTION
**Pull request by mikeldking**
- docs: switch from chroma to weaviate
- switch to also support Cohere
- add a notebook to run evals

### CODE REVIEW

File: python/examples/dspy-rag-fastapi/backend/app/utils/load.py [hunk 1]
```diff
@@ -23,74 +24,58 @@ def chunks(iterable, chunk_size):
...
+        weaviate_client.collections.delete("paul_graham_essay")
+        collection = weaviate_client.collections.create(
+            "paul_graham_essay", vectorizer_config=wvc.config.Configure.Vectorizer.text2vec_openai()
+        )
+
+        logger.info("Loading data into Weaviate.")
+        with collection.batch.dynamic() as batch:
+            for sentence_chunk in chunks(sentence_chunks, 10):
+                embedding_response = openai_client.embeddings.create(
+                    model="text-embedding-ada-002", input=sentence_chunk
```
Comment (axiomofjoy) [inline python/examples/dspy-rag-fastapi/backend/app/utils/load.py:67]: Not sure if it matters, but this is not the latest embedding model.

File: python/examples/dspy-rag-fastapi/README.md [hunk 2]
```diff
@@ -55,6 +55,7 @@ ENVIRONMENT=<your_environment_value>
 INSTRUMENT_DSPY=<true or false>
 COLLECTOR_ENDPOINT=<your_arize_phoenix_endpoint>
 OPENAI_API_KEY=<your_openai_api_key>
+CO_API_KEY=<your_cohere_api_key>
```
Comment (axiomofjoy) [inline python/examples/dspy-rag-fastapi/README.md:58]: Are both necessary, or only one? Would be helpful to clarify.

File: python/examples/dspy-rag-fastapi/backend/Dockerfile [hunk 3]
```diff
@@ -23,8 +23,7 @@ RUN poetry config virtualenvs.create false \
 # Copy the rest of your application's code
 COPY . /rag/
 
-# Populate Chroma db
-# This is done outside of the build
+# Populate Weaviate
 # RUN ["python", "-m", "app.utils.load"]
```
[the remark this answers is not in the document]
Comment (mikeldking) [reply python/examples/dspy-rag-fastapi/backend/Dockerfile:27]: Yeah

File: python/examples/dspy-rag-fastapi/README.md (hunk 2 above)
Comment (mikeldking) [reply python/examples/dspy-rag-fastapi/README.md:58]: I made it semi-optional. You don't get to use cohere if you don't add the key

File: python/examples/dspy-rag-fastapi/backend/app/utils/load.py (hunk 1 above)
Comment (mikeldking) [reply python/examples/dspy-rag-fastapi/backend/app/utils/load.py:67]: yeah weaviate defaults to this. It's a funky setup where the embeddings happen on the server.

File: python/examples/dspy-rag-fastapi/backend/app/utils/load.py [hunk 4]
```diff
@@ -23,74 +24,58 @@ def chunks(iterable, chunk_size):
...
+        openai_client = openai.Client()
+
+        weaviate_client.collections.delete("paul_graham_essay")
+        collection = weaviate_client.collections.create(
+            "paul_graham_essay", vectorizer_config=wvc.config.Configure.Vectorizer.text2vec_openai()
+        )
+
+        logger.info("Loading data into Weaviate.")
+        with collection.batch.dynamic() as batch:
+            for sentence_chunk in chunks(sentence_chunks, 10):
```
[the remark this answers is not in the document]
Comment (mikeldking) [reply python/examples/dspy-rag-fastapi/backend/app/utils/load.py:65]: Will try that

**Review by axiomofjoy — APPROVED** [review]
nice work

File: python/examples/dspy-rag-fastapi/backend/app/utils/rag_functions.py [hunk 5]
```diff
@@ -1,58 +1,99 @@
...
+    """
+    Context manager for WeaviateRM
+    """
+    weaviate_client = weaviate.WeaviateClient(
+        embedded_options=weaviate.embedded.EmbeddedOptions(
+            persistence_data_path=f"{DATA_DIR}/weaviate",
+            additional_env_vars={
+                "ENABLE_MODULES": "text2vec-openai",
+                "DEFAULT_VECTORIZER_MODULE": "text2vec-openai",
+                "OPENAI_APIKEY": os.getenv("OPENAI_API_KEY"),
```
[the remark this answers is not in the document]
Comment (mikeldking) [reply python/examples/dspy-rag-fastapi/backend/app/utils/rag_functions.py:37]: It is

**Pull request merged**